### ISSUE TITLE
Test failure: baseservices/threading/coverage/OSThreadId/osthreadid/osthreadid.sh

### ISSUE THREAD

**Issue by buyaa-n** [dotnet/runtime#73854]
`baseservices/threading/coverage/OSThreadId/osthreadid/osthreadid.sh` failed on https://github.com/dotnet/runtime/pull/73520/checks?check_run_id=7802499930

[Log](https://helix.dot.net/api/2019-06-17/jobs/7a593903-4888-40da-8893-d0ad9dbbbad3/workitems/baseservices.threading/console):
```
baseservices/threading/coverage/OSThreadId/osthreadid/osthreadid.sh [FAIL]
      
      =================================================================
      External Debugger Dump:
      =================================================================
      [New LWP 30683]
      [New LWP 30684]
      [New LWP 30685]
      [New LWP 30689]
      [Thread debugging using libthread_db enabled]
      Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
      warning: the debug information found in "/datadisks/disk1/work/9FB4097F/p/libcoreclr.so.dbg" does not match "/datadisks/disk1/work/9FB4097F/p/libcoreclr.so" (CRC mismatch).
      
      0x00007f7a61ecaad3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7ffc1bee1980) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
      88../sysdeps/unix/sysv/linux/futex-internal.h: No such file or directory.
        Id   Target Id         Frame 
      * 1    Thread 0x7f7a624f6740 (LWP 30679) "corerun" 0x00007f7a61ecaad3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7ffc1bee1980) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
        2    Thread 0x7f7a5e3ff700 (LWP 30683) "SGen worker" 0x00007f7a61ecaad3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f7a61156748) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
        3    Thread 0x7f7a5b67a700 (LWP 30684) "corerun" 0x00007f7a612a1bb9 in __GI___poll (fds=0x7f7a54002d50, nfds=1, timeout=-1) at ../sysdeps/unix/sysv/linux/poll.c:29
        4    Thread 0x7f7a5b479700 (LWP 30685) "Finalizer" 0x00007f7a61ecd7c6 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f7a61147720) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
        5    Thread 0x7f7a5aad8700 (LWP 30689) "corerun" 0x00007f7a61ecf32a in __waitpid (pid=30692, stat_loc=0x7f7a5aad6640, options=0) at ../sysdeps/unix/sysv/linux/waitpid.c:30
      
      Thread 5 (Thread 0x7f7a5aad8700 (LWP 30689)):
      #0  0x00007f7a61ecf32a in __waitpid (pid=30692, stat_loc=0x7f7a5aad6640, options=0) at ../sysdeps/unix/sysv/linux/waitpid.c:30
      #1  0x00007f7a5fb11537 in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #2  0x00007f7a5fab333e in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #3  0x00007f7a5fa0fec1 in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #4  <signal handler called>
      #5  0x00007f7a5f948c72 in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #6  0x00007f7a5c0a4efb in wrapper_managed_to_native_object___icall_wrapper_ves_icall_object_new_specific_intptr () from /datadisks/disk1/work/9FB4097F/p/System.Private.CoreLib.dll.so
      #7  0x00007f7a5b9ac1ff in System_Threading_Thread__get_WaitInfog__AllocateWaitInfo_51_0 () from /datadisks/disk1/work/9FB4097F/p/System.Private.CoreLib.dll.so
      #8  0x00007f7a5e402c58 in ?? ()
      #9  0x00000000ffffffff in ?? ()
      #10 0x0000000000000000 in ?? ()
      
      Thread 4 (Thread 0x7f7a5b479700 (LWP 30685)):
      #0  0x00007f7a61ecd7c6 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x7f7a61147720) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
      #1  do_futex_wait (sem=sem@entry=0x7f7a61147720, abstime=0x0) at sem_waitcommon.c:111
      #2  0x00007f7a61ecd8b8 in __new_sem_wait_slow (sem=0x7f7a61147720, abstime=0x0) at sem_waitcommon.c:181
      #3  0x00007f7a5f97f276 in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #4  0x00007f7a5f957eea in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #5  0x00007f7a5f957d79 in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #6  0x00007f7a61ec46db in start_thread (arg=0x7f7a5b479700) at pthread_create.c:463
      #7  0x00007f7a612ae61f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
      
      Thread 3 (Thread 0x7f7a5b67a700 (LWP 30684)):
      #0  0x00007f7a612a1bb9 in __GI___poll (fds=0x7f7a54002d50, nfds=1, timeout=-1) at ../sysdeps/unix/sysv/linux/poll.c:29
      #1  0x00007f7a5fbd588a in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #2  0x00007f7a5fbd2d95 in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #3  0x00007f7a5fbd14a9 in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #4  0x00007f7a5fbd4ca1 in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #5  0x00007f7a61ec46db in start_thread (arg=0x7f7a5b67a700) at pthread_create.c:463
      #6  0x00007f7a612ae61f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
      
      Thread 2 (Thread 0x7f7a5e3ff700 (LWP 30683)):
      #0  0x00007f7a61ecaad3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7f7a61156748) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
      #1  __pthread_cond_wait_common (abstime=0x0, mutex=0x7f7a611566f8, cond=0x7f7a61156720) at pthread_cond_wait.c:502
      #2  __pthread_cond_wait (cond=0x7f7a61156720, mutex=0x7f7a611566f8) at pthread_cond_wait.c:655
      #3  0x00007f7a5f9f8213 in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #4  0x00007f7a61ec46db in start_thread (arg=0x7f7a5e3ff700) at pthread_create.c:463
      #5  0x00007f7a612ae61f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
      
      Thread 1 (Thread 0x7f7a624f6740 (LWP 30679)):
      #0  0x00007f7a61ecaad3 in futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7ffc1bee1980) at ../sysdeps/unix/sysv/linux/futex-internal.h:88
      #1  __pthread_cond_wait_common (abstime=0x0, mutex=0x7f7a6114c330, cond=0x7ffc1bee1958) at pthread_cond_wait.c:502
      #2  __pthread_cond_wait (cond=0x7ffc1bee1958, mutex=0x7f7a6114c330) at pthread_cond_wait.c:655
      #3  0x00007f7a5f99bc1f in mono_os_event_wait_multiple () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #4  0x00007f7a5f99ba1c in mono_os_event_wait_one () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #5  0x00007f7a5f9ae86a in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #6  0x00007f7a5f953b5e in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #7  0x00007f7a5f910347 in ?? () from /datadisks/disk1/work/9FB4097F/p/libcoreclr.so
      #8  0x00007f7a5bed6d65 in wrapper_managed_to_native_System_Threading_Thread_JoinInternal_System_Threading_Thread_int () from /datadisks/disk1/work/9FB4097F/p/System.Private.CoreLib.dll.so
      #9  0x00007f7a5b9abf48 in System_Threading_Thread_Join () from /datadisks/disk1/work/9FB4097F/p/System.Private.CoreLib.dll.so
      #10 0x00007f7a5e402b88 in ?? ()
      #11 0x00007f7a5aeec35c in Threading_Tests_OSThreadId_Main_string__ () at /__w/1/s/src/tests/baseservices/threading/coverage/OSThreadId/OSThreadId.cs:42
      #12 0x00007ffc1bee1be0 in ?? ()
      #13 0x00007ffc1bee1c80 in ?? ()
      #14 0x000055fc2fb7a3a8 in ?? ()
      #15 0x00007ffc1bee1e98 in ?? ()
      #16 0x000055fc2fc04da0 in ?? ()
      #17 0x00007ffc1bee1bd0 in ?? ()
      #18 0x00007f7a5c09c632 in wrapper_runtime_invoke_object_runtime_invoke_dynamic_intptr_intptr_intptr_intptr () from /datadisks/disk1/work/9FB4097F/p/System.Private.CoreLib.dll.so
      Backtrace stopped: previous frame inner to this frame (corrupt stack?)
      /datadisks/disk1/work/9FB4097F/w/A9F209A2/e/baseservices/threading/coverage/OSThreadId/osthreadid/osthreadid.sh: line 384: 30679 Aborted                 (core dumped) $LAUNCHER $ExePath "${CLRTestExecutionArguments[@]}"
      
      Return code:      1
      Raw output file:      /datadisks/disk1/work/9FB4097F/w/A9F209A2/uploads/Reports/baseservices.threading/coverage/OSThreadId/osthreadid/osthreadid.output.txt
      Raw output:
      BEGIN EXECUTION
      /datadisks/disk1/work/9FB4097F/p/corerun -p System.Reflection.Metadata.MetadataUpdater.IsSupported=false osthreadid.dll ''
      
      =================================================================
      Native Crash Reporting
      =================================================================
      Got a SIGSEGV while executing native code. This usually indicates
      a fatal error in the mono runtime or one of the native libraries 
      used by your application.
      =================================================================
      
      =================================================================
      Native stacktrace:
      =================================================================
      0x7f7a5fb113f2 - Unknown
      0x7f7a5fab333e - Unknown
      0x7f7a5fa0fec1 - Unknown
      0x7f7a61ecf980 - Unknown
      0x7f7a5f948c72 - Unknown
      0x7f7a5c0a4efb - Unknown
      
      =================================================================
      Basic Fault Address Reporting
      =================================================================
      Memory around native instruction pointer (0x7f7a5f948c72):0x7f7a5f948c62  53 48 83 ec 58 48 89 fb c7 44 24 08 00 00 00 00  SH..XH...D$.....
      0x7f7a5f948c72  48 8b 07 48 63 70 1c e8 d2 27 04 00 49 89 c6 48  H..Hcp...'..I..H
      0x7f7a5f948c82  8b 03 c7 44 24 08 00 00 00 00 4d 85 f6 74 11 f6  ...D$.....M..t..
      0x7f7a5f948c92  40 23 20 75 35 4c 89 f0 48 83 c4 58 5b 41 5e c3  @# u5L..H..X[A^.
      
      =================================================================
      Managed Stacktrace:
      =================================================================
        at <unknown> <0xffffffff>
        at System.Object:__icall_wrapper_ves_icall_object_new_specific <0x0006a>
        at System.Threading.Thread:<get_WaitInfo>g__AllocateWaitInfo|51_0 <0x0002e>
        at System.Threading.WaitSubsystem:Wait <0x00046>
        at System.Threading.WaitSubsystem:Wait <0x00036>
        at System.Threading.WaitHandle:WaitOneCore <0x00027>
        at System.Threading.WaitHandle:WaitOneNoCheck <0x0012a>
        at System.Threading.WaitHandle:WaitOne <0x00017>
        at Threading.Tests.OSThreadId:ThreadProc <0x0006b>
        at System.Threading.Thread:StartCallback <0x000db>
        at System.Object:runtime_invoke_dynamic <0x00121>
      =================================================================
      Expected: 100
      Actual: 134
      END EXECUTION - FAILED
      Test Harness Exitcode is : 1
      To run the test:
      > set CORE_ROOT=/datadisks/disk1/work/9FB4097F/p
      > /datadisks/disk1/work/9FB4097F/w/A9F209A2/e/baseservices/threading/coverage/OSThreadId/osthreadid/osthreadid.sh
      Expected: True
      Actual:   False
      Stack Trace:
           at baseservices_threading._coverage_OSThreadId_osthreadid_osthreadid_._coverage_OSThreadId_osthreadid_osthreadid_sh()
           at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
           at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

**Comment by lambdageek** [comment]
Doesn't seem like a race condition in thread waiting.  (saved [a copy of the console log](https://gist.github.com/lambdageek/afbb1500bbc26c98860cf860ef1761da) for reference)

Thread 5 seems to crash while allocating a `ThreadSubsystem.ThreadWaitInfo` (not even calling the constructor; just allocating) possibly the vtable ptr is bad? not sure that would crash, necessarily.  

https://github.com/dotnet/runtime/blob/11e8b1b5c98bfd1b36d326ef7989f94d5083df98/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.ThreadWaitInfo.Unix.cs#L89-L101

The `Thread.WaitInfo` is accessed from the finalizer thread when the thread is exiting, but also from waitable objects (in this case the test has a `ManualResetEvent` when waiting on them).  So it looks like we started a thread and it reached the point of waiting and then crashed.

The main thread (Thread 1) is waiting at the `Thread.Join` loop.  That means `s_resetEvent` is already set.

The allocation of `ThreadWaitInfo` happens on Thread 5 during the wait on the ManualResetEvent which boils down to `WaitSubsystem.Wait(WaitableObject ...)`:

https://github.com/dotnet/runtime/blob/62d1d428ce589caf0dcda593bb24440ce0634b97/src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.Unix.cs#L339

So it's either a codegen bug (maybe we compiled `System_Threading_Thread__get_WaitInfog__AllocateWaitInfo_51_0` wrong?) or a typesystem bug (maybe the vtable is bad)
